src/connection.rs

```diff
@@ -207,4 +217,59 @@ impl Connection {
             },
         }
     }
+
+    fn send_query(&self, tree: &Tree) -> Result<(), Error> {
+        let token = self.query_token.get();
+        // Increment the token for the next request.
+        self.query_token.set(token.wrapping_add(1));
+        let tree = my_try!(json::encode(&tree.to_json()));
+        let len = tree.as_bytes().len();
+        if len > (u32::MAX as usize) {
+            return Err(Error::QueryTooLarge(len));
+        }
+
+        my_try!(self.stream.borrow_mut().write_u64::<LittleEndian>(token));
+        my_try!(self.stream.borrow_mut().write_u32::<LittleEndian>(len as u32));
+        my_try!(self.stream.borrow_mut().write(&tree.as_bytes()));
+
+        Ok(())
+    }
+
+    // TODO: Worry about how to handle unordered responses.  We should probably loop over the
+    // stream once the handshake is done, trying to read each response and putting it in a hashed
+    // collection.
+    fn parse_response_from_stream(stream: &TcpStream) -> Result<QueryResponse, Error> {
+        let mut reader = BufReader::new(stream);
+        let token = my_try!(reader.read_u64::<LittleEndian>());
+        let len = my_try!(reader.read_u32::<LittleEndian>());
+        let mut recv = vec![];
+        let _ = my_try!(reader.take(len as u64).read(&mut recv));
+        let response = my_try!(Json::from_str(my_try!(str::from_utf8(&recv))));
+
+        Ok(QueryResponse {
```

Comment:
The `my_try!` macro will return `Err` explicitly.  I could also have written this

```rust
match Json::from_str(my_try!(str::from_utf8(&recv)) {
  Ok(response) => Ok(QueryResponse {
    query_token: token,
    length:      len,
    response:    response,
  }),
  Err(error) => Err(error),
}
```

Essentially, whenever you see the `my_try!` macro, it's doing error handling on whatever is being called and returning the error if it exists.  So if you reach the last statement in the function, it's guaranteed to be `Ok`.  I may change course though, and use an explicit match statement on the last item returned for the sake of readability.